state/fork/traverse.go

```diff
@@ -0,0 +1,117 @@
+package fork
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// functor that will be called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) error
+
+// TraverseBackward traverses the given fork (specified by block ID `forkHead`)
+// in the order of decreasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
+func TraverseBackward(headers storage.Headers, forkHead flow.Identifier, visitor onVisitBlock, terminal Terminal) error {
+	lowestHeightToVisit, sanityChecker, err := terminal.Translate2Height(headers)
+	if err != nil {
+		return fmt.Errorf("error evaluating terminal: %w", err)
+	}
+
+	startBlock, err := headers.ByBlockID(forkHead)
+	if err != nil {
+		return fmt.Errorf("could not fork head with if  %x: %w", forkHead, err)
+	}
+	if startBlock.Height < lowestHeightToVisit {
+		return nil
+	}
+
+	return unsafeTraverse(headers, startBlock, visitor, lowestHeightToVisit, sanityChecker)
+}
+
+// unsafeTraverse implements the fork traversal in the order of decreasing height.
+// It is unsafe because:
+// * always calls the `visitor` on the `block`, _before_ it checks whether to stop
+// * if `block` has a lower height than lowestHeightToVisit, the traversal keeps
+//   going until we hit the root block
+func unsafeTraverse(headers storage.Headers, block *flow.Header, visitor onVisitBlock, lowestHeightToVisit uint64, sanityChecker sanityCheckLowestVisitedBlock) error {
+	for {
+		err := visitor(block)
+		if err != nil {
+			return err
+		}
+
+		if block.Height == lowestHeightToVisit {
+			err = sanityChecker(block)
```

Comment:
I see the point of having `sanityChecker`, but the `unsafeTraverse` doesn't really need this sanityCheck, having the `unsafeTraverse` taking `sanityChecker` as parameter increases the complexity.

I would suggest to let `unsafeTraverse` return the last visited block, and then the `TraverseForward` and `TraverseBack` could use the sanityChecker to check the last block.

This makes `unsafeTraverse` simplier, and I believe could make `TraverseBackard` easier to understand.

Reply:
another issue is I found the name `sanityCheck` is confusing. I think this is one of the source of increased complexity. 
The question I had when seeing this are: Why do we need sanityCheck? What are these checks? 
I have to navigate to the `Terminal` implementation to understand. 

After reading the sanityChecker, I learned it's to check if all visited blocks are on the same chain. This is a nice check, because if the caller specifies a fork header block and a terminal block to traverse, but what if these two blocks are not on the same chain? It might risk running into an endless loop. 

And this sanity check is only useful if the terminal block is specified by block ID. If it's specified by block height, then there is nothing to sanity check, because we essentially just traverse certain blocks along the chain.

So one suggestion is to just rename `sanityCheck` to `sameChainCheck` to be more specific.